nakamoto-p2p/src/prototype.rs

```diff
@@ -0,0 +1,888 @@
+        pub fn initial_sync(&mut self, peer: PeerId) -> Vec<(PeerId, NetworkMessage)> {
+            // TODO: Notify peer that it should sync.
+            self.state = State::InitialSync(peer);
+
+            let locator_hashes = self.locator_hashes();
+            vec![(peer, self.get_headers(locator_hashes))]
+        }
+
+        /// Check whether or not we are in sync with the network.
+        pub fn is_synced(&self) -> Result<bool, Error> {
+            let height = self.tree.height();
+
+            // TODO: Make sure we only consider connected peers?
+            // TODO: Check actual block hashes once we are caught up on height.
+            if let Some(peer_height) = self.peers.values().map(|p| p.height).min() {
+                Ok(height >= peer_height || peer_height - height <= SYNC_THRESHOLD)
+            } else {
+                Err(Error::NotConnected)
+            }
+        }
+    }
+
+    /// Handshake states.
+    ///
+    /// The steps for an *outbound* handshake are:
+    ///
+    ///   1. Send "version" message.
+    ///   2. Expect "version" message from remote.
+    ///   3. Expect "verack" message from remote.
+    ///   4. Send "verack" message.
+    ///
+    /// The steps for an *inbound* handshake are:
+    ///
+    ///   1. Expect "version" message from remote.
+    ///   2. Send "version" message.
+    ///   3. Send "verack" message.
+    ///   4. Expect "verack" message from remote.
+    ///
+    #[derive(Copy, Clone, Debug, PartialOrd, PartialEq, Ord, Eq)]
+    pub enum Handshake {
+        /// Waiting for "version" message from remote.
+        AwaitingVersion,
+        /// Waiting for "verack" message from remote.
+        AwaitingVerack,
+        /// The peer handshake was completed.
+        Done,
+    }
+
+    impl Default for Handshake {
+        fn default() -> Self {
+            Self::AwaitingVersion
+        }
+    }
+
+    /// Synchronization states.
+    #[derive(Copy, Clone, Debug)]
+    pub enum Synchronize {
+        // TODO: This should keep track of what hash we requested.
+        RequestedHeaders,
+        HeadersReceived,
+        Done,
+    }
+
+    impl Default for Synchronize {
+        fn default() -> Self {
+            Self::RequestedHeaders
+        }
+    }
+
+    /// State of a remote peer.
+    #[derive(Debug)]
+    pub enum PeerState {
+        Handshake(Handshake),
+        Synchronize(Synchronize),
+    }
+
+    /// A remote peer.
+    #[derive(Debug)]
+    pub struct Peer {
+        /// Remote peer address.
+        pub address: net::SocketAddr,
+        /// Local peer address.
+        pub local_address: net::SocketAddr,
+        /// The peer's best height.
+        pub height: Height,
+        /// An offset in seconds, between this peer's clock and ours.
+        /// A positive offset means the peer's clock is ahead of ours.
+        pub time_offset: TimeOffset,
+        /// Whether this is an inbound or outbound peer connection.
+        pub link: Link,
```

Comment:
Could be worthwhile to encode the direction in the peer type, a la:

``` rust
enum Peer {
    Inbound(...),
    Outbound(...),
}
```

Reply:
Yeah, although all fields are shared between the two directions :shrug: